src/Std/Data/DHashMap/Internal/WF.lean

```diff
@@ -135,6 +142,15 @@ theorem pairwise_keys_iff_pairwise_keys [BEq α] [PartialEquivBEq α] {m : Raw 
       (List.keys (toListModel m.buckets)).Pairwise (fun a b => (a == b) = false) :=
   keys_perm_keys_toListModel.pairwise_iff BEq.symm_false
 
+theorem Const.toList_eq_toListModel {β} {m : Raw α (fun _ => β)} :
+    Raw.Const.toList m = Const.toListModel m.buckets := by
+  simp only [Raw.Const.toList, Const.toListModel, ← Raw.toList_eq_toListModel, Raw.toList,
+    Raw.foldRev, Raw.foldRevM, Array.id_run_foldrM, AssocList.foldrM_id]
+  rw [← Array.foldr_hom (List.map _) _ (g₂ := fun l acc => foldr (fun p d => (p.fst, p.snd) :: d) acc l.toList)]
+  · simp
+  · intro l acc
+    induction l <;> simp_all
```

Comment:
This is still ugly, but I think this argument needs to happen *somewhere*, since we have to different `toList` functions (one the double fold consing a `Prod`, the other the double fold consing a `Sigma`).

Reply:
I also dont know any better way of doing it. It is similar for us in #6211 with `insertMany` and `Const.insertMany`

Reply:
Oops, I got confused by the various PRs and added a the relevant comment [here](https://github.com/leanprover/lean4/pull/6304#discussion_r1869478960). As mentioned there, I think having this lemma is good but I would drop the `Const.toListModel` function. You can also optimize the proof using the existing lemmas above:

```lean
theorem foldRev_cons_pair {β} {l : Raw α (fun _ => β)} {acc : List (α × β)} :
    l.foldRev (fun acc k v => (k, v) :: acc) acc =
      (toListModel l.buckets).map (fun ⟨k, v⟩ => (k, v)) ++ acc := by
  simp [foldRev_cons_apply]

theorem Const.toList_eq_toListModel {β} {m : Raw α (fun _ => β)} :
    Raw.Const.toList m = (toListModel m.buckets).map (fun ⟨k, v⟩ => (k, v)) := by
  rw [Raw.Const.toList, foldRev_cons_pair, append_nil]
```